### DISCUSSION TITLE
Supporting tens of thousands of validators on Neatcoin

### DISCUSSION THREAD

**Discussion by sorpaas** [neatcoin/neatcoin#2] [General]
The current usual number of validators BABE and GRANDPA consensus can support (as it is on Kusama) is around 1000. It did not go higher because of other constraints. We do not need to worry about those constraints yet on Neatcoin, so we can experiment getting a much higher number of validators (think of 10000+).

The trick to do this is to realize the fact that the current performance bottleneck is only on GRANDPA, but not BABE. So we can design our staking system to separate the validator set of BABE and GRANDPA. Set GRANDPA's validator set to be still around 200-1000, and set BABE's validator set to be unlimited. In this way we can support (nearly) as many BABE validators as we want without changing the consensus algorithm.

### Design

The concept of nominators ceases to exist. Instead, we have BABE validators and GRANDPA validators.

BABE validators are unlimited. They lock a fixed amount of fund to join in. When they join, they also vote on GRANDPA validators. GRANDPA validators are limited to a fix number (200-1000). They are then elected through the Phragmén method as it happens now. In other words, BABE validators fulfill the current nominator role.

Rewards are split between BABE and GRANDPA validators (the exact number is to be decided, but for example, 70/30). BABE validators are rewarded by their era score. GRANDPA validators are rewarded evenly. Slashing happens as normal -- once slashed, a BABE or GRANDPA validator is forced to be removed from the corresponding validator set.

### Runtime modules

To carry out this design, we need to have Neatcoin's own version of the `pallet-session` and `pallet-staking` module. `pallet-session` will be allowing dedicating BABE and GRANDPA validators separately. `pallet-staking` (which we may rename to `pallet-grandpa-staking`) will only handle GRANDPA validators. A third runtime module `pallet-babe-staking` will handle the BABE validator activation. `pallet-grandpa-staking` and `pallet-babe-staking` share the same era configuration.

### Frontend

Neatcoin also needs to have a separate UI for the new staking modules. Given that we already need this for domain registrations and other stuff, we need this anyway.

### Activation plans

We do not need consensus level changes, but only on runtime level. This makes the activation plan much easier.

Upon activation, we delete current nominator and validator configurations, but keep the current validator set, and use it as the temporary validator set for both BABE and GRANDPA. This temporary validator set will last, for example, for one week. During this period, users can use the new staking module to participate in BABE validation and vote for GRANDPA validators.

**Comment by AlistairStewart** [comment]
GRANDPA should be able to support more validators. You can reduce the average CPU/network load from GRANDPA by increasing the time 2T they wait between rounds of voting. Just increasing this by a factor of 10 would make the load more bursty, but there should be multiple fixes for that. Of course this delays finality and you might prefer not to do that.

If you potentially have a lot of bribable or rationally dishonest BABE validators, you probably want to make sure the block rewards are not too low. On a smart contract chain, block producers might profit from forking the chain to get miner extractible value, i.e. by frontrunning, backrunning and sandwiching all those DeFi transactions that create arbitrage oppurtunities. Creating a 1 block fork risks losing the block reward if the block is not included, but if the MEV for the forked block is much higher than just building one it can be worth the risk.

**Reply by sorpaas** [reply]
> On a smart contract chain, block producers might profit from forking the chain to get miner extractible value, i.e. by frontrunning, backrunning and sandwiching all those DeFi transactions that create arbitrage oppurtunities.

Raising the block reward can only reduce the risk, but not eliminate it, right? The risk exists as well if we only have 200 validators, but just because MEV is low compared with block reward, not many validators would do it.

I do think to deal with the problem in a complete way, we need to do it on runtime level -- make transaction execution order independent from the inclusion order. Record transactions in block A, store it for a while to build enough randomness, then finally execute those transactions in a later block B.

**Comment by burdges** [comment]
Alistair and I originally wanted roughly this mix between smaller GRANDPA pool and larger BABE pool for Polkadot, minus giving BABE block producers any special voting rights, so nominators could only earn block rewards by running a block production node.  We felt it helped on-board new validators and grows the node operator community faster, but then we changed our minds because one validator set simplifies several concern, and parachains could provide similar on-boarding.
